java/server/src/org/openqa/grid/internal/ProxySet.java

```diff
@@ -121,38 +123,23 @@ public boolean isEmpty() {
     return proxies.isEmpty();
   }
 
-  public List<RemoteProxy> getSorted() {
-    List<RemoteProxy> sorted = new ArrayList<>(proxies);
-    Collections.sort(sorted, proxyComparator);
-    return sorted;
-  }
-
-  private Comparator<RemoteProxy> proxyComparator = new Comparator<RemoteProxy>() {
-    @Override
-    public int compare(RemoteProxy o1, RemoteProxy o2) {
-      double p1used = (o1.getTotalUsed() * 1.0) / o1.getTestSlots().size();
-      double p2used = (o2.getTotalUsed() * 1.0) / o2.getTestSlots().size();
-
-      if (p1used == p2used) return 0;
-      return p1used < p2used? -1 : 1;
-    }
-  };
+  public TestSession getNewSession(Map<String, Object> desiredCapabilities) {   
+    // sorting is not working, use a linked list to always put the last used node at the end
+    log.info("Available nodes: " + nextProxyToUse);
 
-  public TestSession getNewSession(Map<String, Object> desiredCapabilities) {
-    // sort the proxies first, by default by total number of
-    // test running, to avoid putting all the load of the first
-    // proxies.
-    List<RemoteProxy> sorted = getSorted();
-    log.info("Available nodes: " + sorted);
-
-    for (RemoteProxy proxy : sorted) {
+    for (RemoteProxy proxy : nextProxyToUse) {
       TestSession session = proxy.getNewSession(desiredCapabilities);
       if (session != null) {
+        synchronized(nextProxyToUse) {
```

Comment:
hi there. concurrent modification exceptions? the changes are atomic to the synchronized operation, so there won't be any concurrent modification (exactly why i have the synchronized block); which is also within the breakout of the loop (the session is not null so will be returned). if there would be a problem it would already have been there in the TestSession session = proxy.getNewSession(desiredCapabilities); line,i'm guessing that must be atomic